File: habit_tracker.go
// habit_tracker.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"time"
	"github.com/google/uuid"
)

type Habit struct {
	ID            string   `json:"id"`
	Name          string   `json:"name"`
	Description   string   `json:"description"`
	Streak        int      `json:"streak"`
	LongestStreak int      `json:"longest_streak"`
	LastLogDate   string   `json:"last_log_date"`
	History       []string `json:"history"`
	CreatedAt     string   `json:"created_at"`
}

func NewHabit(name, desc string) Habit {
	return Habit{
		ID:          uuid.New().String()[:8],
		Name:        name,
		Description: desc,
		Streak:      0,
		History:     []string{},
		CreatedAt:   time.Now().Format(time.RFC3339),
	}
}

func (h *Habit) Log(logDate string) bool {
	if logDate == "" {
		logDate = time.Now().Format("2006-01-02")
	}
	// Check if already logged today
	for _, d := range h.History {
		if d == logDate {
			fmt.Printf("⚠️ Already logged for %s\n", logDate)
			return false
		}
	}
	if h.LastLogDate != "" {
		last, _ := time.Parse("2006-01-02", h.LastLogDate)
		current, _ := time.Parse("2006-01-02", logDate)
		days := int(current.Sub(last).Hours() / 24)
		if days == 1 {
			h.Streak++
		} else if days > 1 {
			h.Streak = 1
		} else if days == 0 {
			// Same day, shouldn't happen due to duplicate check
		}
	} else {
		h.Streak = 1
	}
	if h.Streak > h.LongestStreak {
		h.LongestStreak = h.Streak
	}
	h.LastLogDate = logDate
	h.History = append(h.History, logDate)
	fmt.Printf("✅ Logged '%s' for %s (streak: %d days)\n", h.Name, logDate, h.Streak)
	return true
}

func (h *Habit) Reset() {
	h.Streak = 0
	h.LastLogDate = ""
	fmt.Printf("🔄 Reset streak for '%s'\n", h.Name)
}

type Tracker struct {
	Habits []Habit `json:"habits"`
}

var dataFile = "habits.json"

func (t *Tracker) load() {
	data, err := os.ReadFile(dataFile)
	if err != nil {
		return
	}
	json.Unmarshal(data, t)
}

func (t *Tracker) save() {
	data, _ := json.MarshalIndent(t, "", "  ")
	os.WriteFile(dataFile, data, 0644)
}

func (t *Tracker) getHabit(id string) *Habit {
	for i := range t.Habits {
		if t.Habits[i].ID == id {
			return &t.Habits[i]
		}
	}
	return nil
}

func (t *Tracker) add(name, desc string) {
	h := NewHabit(name, desc)
	t.Habits = append(t.Habits, h)
	t.save()
	fmt.Printf("✅ Habit added: %s (ID: %s)\n", h.Name, h.ID)
}

func (t *Tracker) log(id, date string) {
	h := t.getHabit(id)
	if h == nil {
		fmt.Printf("❌ Habit %s not found.\n", id)
		return
	}
	h.Log(date)
	t.save()
}

func (t *Tracker) list() {
	if len(t.Habits) == 0 {
		fmt.Println("No habits yet. Add one with 'add'")
		return
	}
	fmt.Println("\n🔥 Habit Tracker\n")
	fmt.Println("📋 Your Habits:")
	for i, h := range t.Habits {
		status := "✅"
		if h.Streak == 0 {
			status = "⏳"
		}
		days := "🔥"
		if h.Streak == 0 {
			days = "💤"
		}
		fmt.Printf("  %d. %s %s (%s %d days) – Best: %d days\n", i+1, status, h.Name, days, h.Streak, h.LongestStreak)
	}
}

func (t *Tracker) stats() {
	if len(t.Habits) == 0 {
		fmt.Println("No habits yet.")
		return
	}
	total := len(t.Habits)
	totalLogs := 0
	sumStreak := 0
	longest := 0
	bestHabit := ""
	for _, h := range t.Habits {
		totalLogs += len(h.History)
		sumStreak += h.Streak
		if h.LongestStreak > longest {
			longest = h.LongestStreak
			bestHabit = h.Name
		}
	}
	avg := float64(sumStreak) / float64(total)
	fmt.Println("\n📊 Statistics:")
	fmt.Printf("  Total habits: %d\n", total)
	fmt.Printf("  Total completions: %d\n", totalLogs)
	fmt.Printf("  Average streak: %.1f days\n", avg)
	fmt.Printf("  Longest streak: %d days (%s)\n", longest, bestHabit)
}

func (t *Tracker) reset(id string) {
	h := t.getHabit(id)
	if h == nil {
		fmt.Printf("❌ Habit %s not found.\n", id)
		return
	}
	h.Reset()
	t.save()
}

func (t *Tracker) history(id string) {
	h := t.getHabit(id)
	if h == nil {
		fmt.Printf("❌ Habit %s not found.\n", id)
		return
	}
	fmt.Printf("\n📜 History for '%s':\n", h.Name)
	fmt.Printf("  Current streak: %d days\n", h.Streak)
	fmt.Printf("  Longest streak: %d days\n", h.LongestStreak)
	if len(h.History) > 0 {
		fmt.Println("  Recent logs:")
		start := 0
		if len(h.History) > 10 {
			start = len(h.History) - 10
		}
		for _, d := range h.History[start:] {
			fmt.Printf("    - %s\n", d)
		}
	} else {
		fmt.Println("  No logs yet.")
	}
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: habit_tracker <command> [options]")
		return
	}
	t := &Tracker{}
	t.load()
	cmd := os.Args[1]

	switch cmd {
	case "add":
		addCmd := flag.NewFlagSet("add", flag.ExitOnError)
		name := addCmd.String("name", "", "")
		desc := addCmd.String("desc", "", "")
		addCmd.Parse(os.Args[2:])
		if *name == "" && len(addCmd.Args()) > 0 {
			*name = addCmd.Args()[0]
		}
		if *name == "" {
			fmt.Println("add requires a name")
			return
		}
		t.add(*name, *desc)

	case "log":
		logCmd := flag.NewFlagSet("log", flag.ExitOnError)
		id := logCmd.String("id", "", "")
		date := logCmd.String("date", "", "")
		logCmd.Parse(os.Args[2:])
		if *id == "" && len(logCmd.Args()) > 0 {
			*id = logCmd.Args()[0]
		}
		if *id == "" {
			fmt.Println("log requires an ID")
			return
		}
		t.log(*id, *date)

	case "list":
		t.list()

	case "stats":
		t.stats()

	case "reset":
		if len(os.Args) < 3 {
			fmt.Println("reset <id>")
			return
		}
		t.reset(os.Args[2])

	case "history":
		if len(os.Args) < 3 {
			fmt.Println("history <id>")
			return
		}
		t.history(os.Args[2])

	default:
		fmt.Println("Unknown command.")
	}
}

File: habit_tracker.py
# habit_tracker.py
import json
import os
import sys
import argparse
import uuid
from datetime import datetime, timedelta, date

DATA_FILE = "habits.json"

class Habit:
    def __init__(self, name, description="", habit_id=None):
        self.id = habit_id or str(uuid.uuid4())[:8]
        self.name = name
        self.description = description
        self.streak = 0
        self.longest_streak = 0
        self.last_log_date = None
        self.history = []  # list of ISO date strings
        self.created_at = datetime.now().isoformat()

    def to_dict(self):
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "streak": self.streak,
            "longest_streak": self.longest_streak,
            "last_log_date": self.last_log_date,
            "history": self.history,
            "created_at": self.created_at
        }

    @classmethod
    def from_dict(cls, data):
        habit = cls(data["name"], data.get("description", ""), data.get("id"))
        habit.streak = data.get("streak", 0)
        habit.longest_streak = data.get("longest_streak", 0)
        habit.last_log_date = data.get("last_log_date")
        habit.history = data.get("history", [])
        habit.created_at = data.get("created_at", datetime.now().isoformat())
        return habit

    def log(self, log_date=None):
        if log_date is None:
            log_date = date.today().isoformat()
        # Check if already logged today
        if log_date in self.history:
            print(f"⚠️ Already logged for {log_date}")
            return False
        # Check if this is a continuation or restart
        if self.last_log_date:
            last = datetime.fromisoformat(self.last_log_date).date()
            current = datetime.fromisoformat(log_date).date()
            # If logged yesterday or earlier in the same day, increment streak
            if (current - last).days == 1:
                self.streak += 1
            elif (current - last).days > 1:
                # Gap detected, reset streak
                self.streak = 1
            elif (current - last).days == 0:
                # Same day, should have been caught above
                self.streak = self.streak
        else:
            # First time logging
            self.streak = 1
        # Update longest streak
        if self.streak > self.longest_streak:
            self.longest_streak = self.streak
        self.last_log_date = log_date
        self.history.append(log_date)
        print(f"✅ Logged '{self.name}' for {log_date} (streak: {self.streak} days)")
        return True

    def reset(self):
        self.streak = 0
        self.last_log_date = None
        print(f"🔄 Reset streak for '{self.name}'")

class HabitTracker:
    def __init__(self):
        self.habits = []
        self.load()

    def load(self):
        if os.path.exists(DATA_FILE):
            with open(DATA_FILE, "r") as f:
                data = json.load(f)
                self.habits = [Habit.from_dict(h) for h in data]

    def save(self):
        with open(DATA_FILE, "w") as f:
            json.dump([h.to_dict() for h in self.habits], f, indent=2)

    def get_habit(self, habit_id):
        for h in self.habits:
            if h.id == habit_id:
                return h
        return None

    def add(self, name, description=""):
        habit = Habit(name, description)
        self.habits.append(habit)
        self.save()
        print(f"✅ Habit added: {habit.name} (ID: {habit.id})")

    def log(self, habit_id, log_date=None):
        habit = self.get_habit(habit_id)
        if not habit:
            print(f"❌ Habit {habit_id} not found.")
            return
        habit.log(log_date)
        self.save()

    def list(self):
        if not self.habits:
            print("No habits yet. Add one with 'add'")
            return
        print("\n🔥 Habit Tracker\n")
        print("📋 Your Habits:")
        for i, h in enumerate(self.habits, 1):
            status = "✅" if h.streak > 0 else "⏳"
            days = "🔥" if h.streak > 0 else "💤"
            print(f"  {i}. {status} {h.name} ({days} {h.streak} days) – Best: {h.longest_streak} days")

    def stats(self):
        if not self.habits:
            print("No habits yet.")
            return
        total = len(self.habits)
        total_logs = sum(len(h.history) for h in self.habits)
        avg_streak = sum(h.streak for h in self.habits) / total
        longest = max(h.longest_streak for h in self.habits)
        best_habit = max(self.habits, key=lambda h: h.longest_streak)
        print("\n📊 Statistics:")
        print(f"  Total habits: {total}")
        print(f"  Total completions: {total_logs}")
        print(f"  Average streak: {avg_streak:.1f} days")
        print(f"  Longest streak: {longest} days ({best_habit.name})")

    def reset(self, habit_id):
        habit = self.get_habit(habit_id)
        if not habit:
            print(f"❌ Habit {habit_id} not found.")
            return
        habit.reset()
        self.save()

    def history(self, habit_id):
        habit = self.get_habit(habit_id)
        if not habit:
            print(f"❌ Habit {habit_id} not found.")
            return
        print(f"\n📜 History for '{habit.name}':")
        print(f"  Current streak: {habit.streak} days")
        print(f"  Longest streak: {habit.longest_streak} days")
        if habit.history:
            print("  Recent logs:")
            for log_date in habit.history[-10:]:
                print(f"    - {log_date}")
        else:
            print("  No logs yet.")

def main():
    parser = argparse.ArgumentParser(description="Habit Tracker")
    subparsers = parser.add_subparsers(dest="cmd", required=True)

    add_parser = subparsers.add_parser("add")
    add_parser.add_argument("name")
    add_parser.add_argument("--desc", default="")

    log_parser = subparsers.add_parser("log")
    log_parser.add_argument("habit_id")
    log_parser.add_argument("--date", help="YYYY-MM-DD")

    subparsers.add_parser("list")
    subparsers.add_parser("stats")

    reset_parser = subparsers.add_parser("reset")
    reset_parser.add_argument("habit_id")

    history_parser = subparsers.add_parser("history")
    history_parser.add_argument("habit_id")

    args = parser.parse_args()
    tracker = HabitTracker()

    if args.cmd == "add":
        tracker.add(args.name, args.desc)
    elif args.cmd == "log":
        tracker.log(args.habit_id, args.date)
    elif args.cmd == "list":
        tracker.list()
    elif args.cmd == "stats":
        tracker.stats()
    elif args.cmd == "reset":
        tracker.reset(args.habit_id)
    elif args.cmd == "history":
        tracker.history(args.habit_id)

if __name__ == "__main__":
    main()
